safehttp/plugins/xsrf/angular/xsrf.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xsrfangular
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"github.com/google/go-safeweb/safehttp"
+	"github.com/google/go-safeweb/safehttp/plugins/xsrf"
+)
+
+// Interceptor provides protection against Cross-Site Request Forgery attacks
+// for Angular's XHR requests.
+//
+// See https://docs.angularjs.org/api/ng/service/$http#cross-site-request-forgery-xsrf-protection for more details.
+type Interceptor struct {
+	// TokenCookieName is the name of the seesion cookie that holds the XSRF
+	// token.
+	TokenCookieName string
+	// TokenHeaderName is the name of the HTTP header that holds the XSRF token.
+	TokenHeaderName string
+}
+
+var _ safehttp.Interceptor = &Interceptor{}
+
+// Default creates an Interceptor with TokenCookieName set to XSRF-TOKEN and
+// TokenHeaderName set to X-XSRF-TOKEN, their default values. However, in order
+// to prevent collisions when multiple applications share the same domain or
+// subdomain, each application should set a unique name for the cookie.
+func Default() *Interceptor {
+	return &Interceptor{
+		TokenCookieName: "XSRF-TOKEN",
+		TokenHeaderName: "X-XSRF-TOKEN",
+	}
+}
+
+// Before checks for the presence of a matching XSRF token, generated on the
+// first page access, in both a cookie and a header. Their names should be set
+// when the Interceptor is created.
+func (it *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, _ safehttp.InterceptorConfig) safehttp.Result {
+	if xsrf.StatePreserving(r) {
+		return safehttp.NotWritten()
+	}
+
+	c, err := r.Cookie(it.TokenCookieName)
+	if err != nil || c.Value() == "" {
+		return w.WriteError(safehttp.StatusForbidden)
+	}
+
+	tok := r.Header.Get(it.TokenHeaderName)
+	if tok == "" || tok != c.Value() {
+		// Only JavaScript running on the user domain can read the
```

Comment:
user domain?

Reply:
user of the framework domain, should I rephrase it?

Reply:
Be consistent with the phrasing from https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies.

You can say that JavaScript has access to cookies from the same domain as it's running on.

Introducing a "user" is confusing here.